### PR TITLE
ISSUE #2796 - Containers hook-up

### LINKED ISSUE #2796: Revamping - hook up with containers
### Description
We need to get real data for the containers.

### Goals
- [x] All containers will be filled with all the containers for that selected project
- [x] Favourite containers will be filled with the data from the real favourite containers for the selected project

### Tasks
- [x] Add api endpoint */teamspaces/:teamspace/projects/:projectId/containers* in service
- [x] Add redux/saga action for fetching the containers
- [x] Add tests for fetching the containers as depicted in *test\sagas.spec.ts*
- [x] Add selector for all containers
- [x] Add selector for favourites containers
- [x] Trigger the fetching of this data when loading containers. If there were already data loaded dont refetch it.
- [x] Use hook creators (as depicted in *src\v5\services\selectorsHooks\currentUserSelectors.hooks.ts*) and use those in the containers page.
- [x] Show loading feedback when loading the containers

### Related Resources
MockUp : `https://www.figma.com/file/djgxvRKOaMMyutswn5yywU/Teamspace-(live)?node-id=821%3A1258

### PR DESCRIPTION
**Pull request by EmilTheSadCat**
This fixes #2796

#### Description
- Handle fetching containers list and stats
- Add skeleton block and spinner loaders

test URL: https://issue-2796.dev.3drepo.io/v5/dashboard/carmen/843ba9f0-00c6-11ec-8d97-41a278fb55fd/t/containers

### CODE REVIEW

**Comment by carmenfan** [comment]
@EmilTheSadCat another super weird thing is how it tells you it has no containers until we know how many items there are. Can we get a loading state for that? @george-hadfield

**Comment by george-hadfield** [comment]
> @EmilTheSadCat I think we need a placeholder for the project name while it's trying to find out what is the project name, as it is super weird for the breadcrumb to autopopulate like that: aa
> 
> Maybe just put the project ID as a placeholder until we know? tagging @george-hadfield incase he has anything in mind.

I would simply have "Loading project" in place until we have the project name to replace it.

**Comment by george-hadfield** [comment]
> @carmenfan @george-hadfield
> 
> ### Latest revision
> > image
> > Can we put the date/time instead of the revision ID if the tag is not available?
> 
> Sure, we can, but this is what we get in `latestRevision` field from API. We can check for the uuid or the length on FE, but I'm afraid it won't be reliable. Perhaps BE shouldn't return this field at all if there is no revision name.
> 
> Another thing is - how should we handle displaying long containers or revision names? They may destroy the UI if too long. It may apply to other views as well. I think we need some design solution for this.
> 
> ### Skeleton loaders
> > > > @EmilTheSadCat another super weird thing is how it tells you it has no containers until we know how many items there are. Can we get a loading state for that? @george-hadfield
> > > 
> > > 
> > > The idea was that it would not show the empty states before the skeleton screen. The skeleton screen was intended to be in place whilst all of the content loads. Check out the skeleton on https://www.youtube.com/ and you'll see what I mean.
> > 
> > 
> > @EmilTheSadCat @george-hadfield In that case could we show the skeleton screen with a fake number of items? to avoid the empty state, before we know the real number of items - as this is the problem here i think.
> > Otherwise we need another skeleton screen for before we know the number of items.
> 
> We can do fake list, but I'd like to know how many items should be displayed then? My thought though is that it can be confusing for the user to see his 3-4 items loading, but after loading is done he has none. Another skeleton screen for before we know the number of items may be a solution to this.

Could we use a gradient from the left of the revisions label to hide the extra characters as shown in the below design. Then if a user hovers on the latest revision text, it displays the full revision name?

image
https://www.figma.com/file/djgxvRKOaMMyutswn5yywU/Teamspace-live?node-id=2246%3A33202

**Comment by EmilTheSadCat** [comment]
@carmenfan @george-hadfield 
Hi, this is ready for another review.

> It looks like the error tooltip is still using dummy data. can we update this?

Error tooltip is using real data now. 

> Another thing I've noticed is that the UI just hangs if the endpoints failed. We should be showing the pop up error message to indicate to the user that something went wrong

This is already implemented. I've replicated it using the way you proposed and it's showing error popup as intended.

@george-hadfield  Regarding text overflow solution I've created a short vid to show how it works at the moment.

https://user-images.githubusercontent.com/39130679/139840653-f452e8ad-5cfa-4742-a1af-9a2a33d7db0a.mov

**Comment by george-hadfield** [comment]
> @carmenfan @george-hadfield Hi, this is ready for another review.
> 
> > It looks like the error tooltip is still using dummy data. can we update this?
> 
> Error tooltip is using real data now.
> 
> > Another thing I've noticed is that the UI just hangs if the endpoints failed. We should be showing the pop up error message to indicate to the user that something went wrong
> 
> This is already implemented. I've replicated it using the way you proposed and it's showing error popup as intended.
> 
> @george-hadfield Regarding text overflow solution I've created a short vid to show how it works at the moment.
> 
>  text-overflow-example.mov

This works great thanks @EmilTheSadCat

**Comment by george-hadfield** [comment]
> Also, (probably more for @george-hadfield ), should we have a placeholder for `latest revision` text if the there is 0 revision? it looks a bit odd to have it empty like this. image

If there are no revisions (empty container), then can we replaces the "Latest revision" text with "Container empty"

File: frontend/src/v5/ui/routes/dashboard/projects/containers/containersList/latestRevision/latestRevision.component.tsx [hunk 1]
```diff
@@ -16,96 +16,34 @@
...
+			</Label>
+		) : (
+			<>
+				<Label>
+					<Trans
+						id="containers.list.item.latestRevision.label"
+						message="Latest revision: "
+					/>
+				</Label>
+				<RevisionStatus {...props} />
```
Comment (carmenfan) [inline frontend/src/v5/ui/routes/dashboard/projects/containers/containersList/latestRevision/latestRevision.component.tsx:45]: @EmilTheSadCat 

I think this needs to be outside of the if statement? It's possible that we have no revision, but currently uploading one.

File: frontend/src/v5/ui/routes/dashboard/projects/containers/containersList/latestRevision/revisionStatus/revisionStatus.component.tsx [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+	name: string;
+	status: ContainerStatuses;
+	error?: {
+		date: Date | null;
+		message: string;
+	}
+}
+
+export const RevisionStatus = ({ status, error, name }: IRevisionStatus): JSX.Element => {
+	if (status === ContainerStatuses.QUEUED) {
```
[the remark this answers is not in the document]
Comment (EmilTheSadCat) [reply frontend/src/v5/ui/routes/dashboard/projects/containers/containersList/latestRevision/revisionStatus/revisionStatus.component.tsx:36]: I think that it's more straightforward using if statements. The default case is handled at the bottom of the component. It's a common practice in React.

> Also there are 2 missing status - Queued for Unity and Generating Bundles.

Can you please write down the exact strings to be expected for a `status` field in a response for those statuses? I couldn't find them in the system.

[the remark this answers is not in the document]
Comment (carmenfan) [reply frontend/src/v5/ui/routes/dashboard/projects/containers/containersList/latestRevision/revisionStatus/revisionStatus.component.tsx:36]: > I think that it's more straightforward using if statements. The default case is handled at the bottom of the component. It's a common practice in React.

I really don't think this has anything to do with React - this is just basic coding of conditions. Imo switch case would make it clearer what you're doing, but if you want to stick with if then that's fine.

> Can you please write down the exact strings to be expected for a status field in a response for those statuses? I couldn't find them in the system.

there is no comprehensive list anywhere:
- ok (or undefined also means ok).
- failed
and some form of processing: https://github.com/3drepo/3drepobouncer/blob/master/tools/bouncer_worker/src/constants/statuses.js

**Review by george-hadfield — DISMISSED** [review]
All looks fine to me!

**Comment by carmenfan** [comment]
@EmilTheSadCat  I just merged in containers revisions and this caused conflicts here - can you resolve them please?

**Comment by m1chalkubiak** [comment]
> @EmilTheSadCat I just merged in containers revisions and this caused conflicts here - can you resolve them please?

@carmenfan As Emil is off today and he will not be available until next Monday, I will deal with these conflicts. The changes conflict directly with the ones I made, so I should be able to do it correctly.

**Comment by carmenfan** [comment]
> > @EmilTheSadCat I just merged in containers revisions and this caused conflicts here - can you resolve them please?
> 
> @carmenfan As Emil is off today and he will not be available until next Monday, I will deal with these conflicts. The changes conflict directly with the ones I made, so I should be able to do it correctly.

thanks @m1chalkubiak !

**Comment by carmenfan** [comment]
@m1chalkubiak  

I just had a quick play, it seems to be fine.

I think in the interest of keeping everything as aligned as possible we'll merge this in now and if you can work on the rev details on the v2 containers branch that'd be good.

**Pull request merged**